File: app/compiler.py
MAX_SIZE = 5e6
MAX_CELLS  = 65536
MAX_VAL = 255
MIN_VAL = 0

lineno = 1
ptr = 0
data = [0] * MAX_CELLS
brackets = []

inp_buff = ''
out_buff = ''
inp_ind = 0

def interpret(program,inputPath,outputPath,Q) :
	global MAX_CELLS, MAX_VAL, MIN_VAL, MAX_SIZE
	global lineno, ptr, data, brackets, inp_buff, out_buff, inp_ind

	strlen = len(program)
	if (strlen > MAX_SIZE) :
		Q.put("MEMORY LIMIT EXCEEDED")
		return

	with open(inputPath,'r') as f :
		inp_buff = f.read()
	inp_buff = inp_buff.strip()
	inplen = len(inp_buff)

	ind = 0
	while (ind < strlen) :
		if (program[ind] == '+') :
			if (data[ptr] >= MAX_VAL) :
				data[ptr] = MIN_VAL
			else :
				data[ptr] += 1
		elif (program[ind] == '-') :
			if (data[ptr] <= MIN_VAL) :
				data[ptr] = MAX_VAL
			else :
				data[ptr] -= 1
		elif (program[ind] == '<') :
			ptr -= 1
			if (ptr < 0) :
				ptr = MAX_CELLS-1
		elif (program[ind] == '>') :
			ptr += 1
			if (ptr >= MAX_CELLS) :
				ptr = 0
		elif (program[ind] == ',') :
			if (inp_ind >= inplen) :
				Q.put(f"RUNTIME ERROR : Line {lineno}")
				return
			data[ptr] = ord(inp_buff[inp_ind])
			inp_ind += 1
		elif (program[ind] == '.') :
			out_buff += str(chr(data[ptr]))
		elif (program[ind] == '[') :
			if (data[ptr] != 0) :
				brackets.append([ind,lineno])
			else :
				ob = 1
				cb = 0
				ind += 1
				while (ind < strlen) :
					if (program[ind] == '[') :
						ob += 1
					elif (program[ind] == ']') :
						cb += 1
					elif (program[ind] == '\n') :
						lineno += 1
					if (cb == ob) :
						break
					ind += 1

				if(ob != cb) :
					Q.put(f"SYNTAX ERROR : Line {lineno}")
					return
		elif (program[ind] == ']') :
			if (len(brackets) == 0) :
				Q.put(f"SYNTAX ERROR : Line {lineno}")
				return
			if (data[ptr] == 0) :
				brackets.pop()
			else :
				lineno = brackets[-1][1]
				ind = brackets[-1][0]
		elif (program[ind] == '\n') :
			lineno += 1
		ind += 1

	if (len(brackets) != 0) :
		Q.put(f"SYNTAX ERROR : Line {lineno}")
		return

	with open(outputPath,'w+') as f :
		f.write(out_buff)

	Q.put("ANSWER WRITTEN")
	return

File: app/evaluation/compiler/compiler.cpp
#include<iostream>
#include<stack>
#include<string.h>
#include<utility>
#include<fstream>

#define MAX_SIZE 1e6
#define MAX_CELLS 65536
#define MAX_VAL 255
#define MIN_VAL 0

using namespace std;

int lineno = 1;
unsigned char data[MAX_CELLS] = {0};
int ptr = 0;
stack<pair<int,int> > brackets;

string inp_buff = "",out_buff = "";
int inp_ind = 0;

int interpret(string &program,int start = 0){
	int ind = start;
	while(program[ind] != EOF){
		if(program[ind] == '+'){
			if(data[ptr] >= MAX_VAL) data[ptr] = MIN_VAL;
			else data[ptr]++;
		}
		else if(program[ind] == '-'){
			if(data[ptr] <= MIN_VAL) data[ptr] = MAX_VAL;
			else data[ptr]--;
		}
		else if(program[ind] == '<'){
			ptr--;
			if(ptr < 0) ptr = MAX_CELLS-1;
		}
		else if(program[ind] == '>'){
			ptr++;
			if(ptr >= MAX_CELLS) ptr = 0;
		}
		else if(program[ind] == ','){
			if(inp_ind >= inp_buff.length()){
				cout<<"RUNTIME ERROR : No Input Character for Line "<<lineno<<endl;
				return 0;
			}
			data[ptr] = inp_buff[inp_ind++];
		}
		else if(program[ind] == '.'){
			out_buff += data[ptr];
		}
		else if(program[ind] == '['){
			if(data[ptr] != 0){
				brackets.push(make_pair(ind,lineno));
				ind++;
				continue;
			}
			int ob = 1,cb = 0,p_lineno = lineno;
			while(program[++ind] != EOF){
				if(program[ind] == '[') ob++;
				else if(program[ind] == ']') cb++;
				else if(program[ind] == '\n') lineno++;
				if(cb == ob) break;
			}
			if(ob != cb){
				cerr<<"SYNTAX ERROR : Corresponding Closing Brackets not found for Line "<<p_lineno<<endl;
				return 0;
			}
		}
		else if(program[ind] == ']'){
			if(brackets.size() == 0){
				cerr<<"SYNTAX ERROR : Corresponding Opening Brackets not found for Line "<<lineno<<endl;
				return 0;
			}
			if(data[ptr] == 0){
				brackets.pop();
				ind++;
				continue;
			}
			lineno = brackets.top().second;
			ind = brackets.top().first;
		}
		else if(program[ind] == '\n'){
			lineno++;
		}
		ind++;
	}
	if(brackets.size()) {
		cerr<<"SYNTAX ERROR : Brackets not matching"<<endl;
		return 0;
	}
	return 0;
}

int main(int argc,char **argv){
	string input(argv[2]),output(argv[3]);
	string source(argv[1]);

	ifstream fsource(source.c_str());

	ifstream fin(input.c_str());
	ofstream fout(output.c_str());

	if(fin == NULL || fout == NULL || fsource == NULL){
		cout<<"FILE ERROR : Error while opening files"<<endl;
		return 0;
	}

	string program = "";
	string tmp;
	long long noi = 0;
	while(getline(fsource,tmp)){
		noi += tmp.length();
		program += tmp;
		program += "\n";
	}
	program += EOF;

	if(noi > MAX_SIZE){
		cout<<"MEMORY ERROR : Number of Instructions exceeded the Maximum Limit"<<endl;
		return 0;
	}

	while(fin>>tmp){
		inp_buff += tmp;
	}

	interpret(program);

	//out_buff += EOF;
	fout.write(out_buff.c_str(),out_buff.length());

	fin.close();
	fout.close();
	fsource.close();

	return 0;
}